### PR TITLE
Adding feature bucket documentation and generated enum

### PR DESCRIPTION
**Pull request by grafnu**
Also renames the code use of "category" to "buckets" to avoid confusion with the already existing error categories.

### CODE REVIEW

File: bin/gencode_buckets [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+WILDCARD_SUFFIX = r'.[a-z]+(_[a-z]+)*'
+
+BUCKET_REGEX = r' *\* _([a-z]+)_: (.*)'
+JSON_FORMAT = '%s%s{ "pattern": "^%s$" }'
+DEVICE_PREFIX = 'device'
+
+JAVA_DESCRIPTION = '\n%s// %s\n'
+JAVA_TARGET = '%s%s("%s")%s\n'
+
+doc_in = os.path.join('docs/specs/buckets.md')
```
Comment (johnrandolph) [inline bin/gencode_buckets:18]: Make doc_in, java_in, java_out CONSTANTS, or, move them to main() and pass them as arguments to the functions (better for testing too). Prefer latter

Also, not sure on the need for os.path.join() ?

File: bin/gencode_buckets [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+def read_buckets():
+  buckets = []
+  prefix = []
+  previous = -1
+  group = None
+  with open(doc_in, 'r', encoding='utf-8') as doc:
+    while line := doc.readline():
+      indent = line.find(BUCKET_MARKER)//2
+      wildcard = re.match(WILDCARD_REGEX, line)
+      if wildcard:
```
Comment (johnrandolph) [inline bin/gencode_buckets:31]: Where is this used?

**Review by johnrandolph — COMMENTED** [review]
Since this script will write java code,

File: bin/gencode_buckets (hunk 2 above)
Comment (grafnu) [reply bin/gencode_buckets:31]: It's not -- good catch (removed).

File: bin/gencode_buckets (hunk 1 above)
Comment (grafnu) [reply bin/gencode_buckets:18]: Did both actually (since they're CONSTANTS that should be passed in). And removed os.path.join()

**Review by johnrandolph — APPROVED** [review]

**Pull request merged**